src/ConfigTemplateDefinitionResolver.php

```diff
@@ -0,0 +1,435 @@
+use Drupal\Core\Field\FieldDefinitionInterface;
+use Drupal\update_helper\ConfigName;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Config template definition resolver.
+ *
+ * @package Drupal\testsite_builder
+ */
+class ConfigTemplateDefinitionResolver {
+
+  /**
+   * The config template definition.
+   *
+   * @var \Drupal\testsite_builder\ConfigTemplateDefinition
+   */
+  protected $templateDefinition;
+
+  /**
+   * The config template type manager service.
+   *
+   * @var \Drupal\testsite_builder\ConfigTemplateTypePluginManager
+   */
+  protected $configTemplateTypeManager;
+
+  /**
+   * The entity type manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * The field manager service.
+   *
+   * @var \Drupal\Core\Entity\EntityFieldManagerInterface
+   */
+  protected $fieldManager;
+
+  /**
+   * The queue of field configurations what should be created.
+   *
+   * @var array
+   */
+  protected $fieldTemplatesQueue = [];
+
+  /**
+   * The configuration template collection.
+   *
+   * @var \Drupal\testsite_builder\ConfigTemplateDefinitionCollection
+   */
+  protected $collection;
+
+  /**
+   * The entity type.
+   *
+   * @var string
+   */
+  protected $entityType;
+
+  /**
+   * The config template definition constructor.
+   *
+   * @param \Drupal\testsite_builder\ConfigTemplateTypePluginManager $config_template_type_manager
+   *   The config template type manager service.
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   The entity type manager service.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $field_manager
+   *   The field manager service.
+   */
+  protected function __construct(ConfigTemplateTypePluginManager $config_template_type_manager, EntityTypeManagerInterface $entity_type_manager, EntityFieldManagerInterface $field_manager) {
+    $this->configTemplateTypeManager = $config_template_type_manager;
+    $this->entityTypeManager = $entity_type_manager;
+    $this->fieldManager = $field_manager;
+  }
+
+  /**
+   * Creates instance of config template definition.
+   *
+   * @param \Symfony\Component\DependencyInjection\ContainerInterface $container
+   *   The symfony container.
+   * @param \Drupal\testsite_builder\ConfigTemplateDefinitionCollection $collection
+   *   The config template collection.
+   *
+   * @return \Drupal\testsite_builder\ConfigTemplateDefinitionResolver
+   *   The instance of config template definition resolver.
+   */
+  public static function create(ContainerInterface $container, ConfigTemplateDefinitionCollection $collection): ConfigTemplateDefinitionResolver {
+    $instance = new static(
+      $container->get('testsite_builder.config_template_type_manager'),
+      $container->get('entity_type.manager'),
+      $container->get('entity_field.manager')
+    );
+
+    $instance->collection = $collection;
```

Comment:
I might be missing something here, but this creation pattern makes only sense, if you either create the $instance from a parent class, as you do in Drupal\testsite_builder\Plugin\testsite_builder\EntityType\Media::create(). Or when you have multiple factories, as in ConfigTemplateDefinitionCollection::createFromTemplates().

In this case, you can just add the $collection to the __construct parameters and handle it there.

Reply:
fixed in #37